src/analyzer.ts

```diff
@@ -82,102 +90,117 @@ export class Analyzer {
+  /**
+   * Like `analyze`, but for use cases like editors where the caller believes
+   * the file has changed since the last time it was analyzed, and you may
+   * even have unsaved in-memory contents to use for the given file.
+   */
+  async analyzeChangedFile(url: string, contents?: string):
+      Promise<Document|null> {
+    const resolved = this._resolveUrl(url);
+    this._scannedDocuments.delete(resolved);
+    this._parsedDocuments.delete(resolved);
+    const scannedDocument =
+        await this._analyzeResolved(this._resolveUrl(url), contents);
+    return Document.makeRootDocument(scannedDocument);
+  }
+
+  private async _analyzeResolved(resolvedUrl: string, contents?: string):
+      Promise<ScannedDocument|null> {
+    let cachedResult = this._scannedDocuments.get(resolvedUrl);
     if (cachedResult) {
       return cachedResult;
     }
     let promise = (async() => {
       // Make sure we wait and return a Promise before doing any work, so that
-      // the Promise can be cached.
+      // the Promise is cached before anything else happens.
       await Promise.resolve();
-      let document = await this._loadResolved(resolvedUrl);
+      let document = await this._loadResolved(resolvedUrl, contents);
       return this._analyzeDocument(document);
     })();
-    this._documentDescriptors.set(resolvedUrl, promise);
+    this._scannedDocuments.set(resolvedUrl, promise);
     return promise;
   }
 
-  async resolve(): Promise<Analysis> {
-    return new Analysis(await Promise.all(this._documentDescriptors.values()));
+  private async _resolve(url: string): Promise<ScannedDocument|null> {
+    return this._analyzeResolved(this._resolveUrl(url));
   }
 
   /**
    * Parses and analyzes a document from source.
    */
   private async _analyzeSource(
       type: string, contents: string, url: string,
-      locationOffset?: LocationOffset): Promise<DocumentDescriptor> {
+      locationOffset?: LocationOffset,
+      attachedComment?: string): Promise<ScannedDocument> {
     let resolvedUrl = this._resolveUrl(url);
     let document = this._parse(type, contents, resolvedUrl);
-    return await this._analyzeDocument(document, locationOffset);
+    return await this._analyzeDocument(
+        document, locationOffset, attachedComment);
   }
 
   /**
    * Analyzes a parsed Document object.
    */
   private async _analyzeDocument(
-      document: Document<any, any>,
-      maybeLocationOffset?: LocationOffset): Promise<DocumentDescriptor> {
+      document: ParsedDocument<any, any>, maybeLocationOffset?: LocationOffset,
+      maybeAttachedComment?: string): Promise<ScannedDocument> {
     const locationOffset =
         maybeLocationOffset || {line: 0, col: 0, filename: document.url};
-    let entities = await this.getEntities(document);
+    let entities = await this._getEntities(document);
     for (const entity of entities) {
-      if (entity instanceof ElementDescriptor) {
+      if (entity instanceof ScannedElement) {
```

Comment:
Agree. I'd like every scanned feature to require a source location, and to handle location offsets